lib/kafka_ex/partitioner.ex

```diff
@@ -0,0 +1,43 @@
+defmodule KafkaEx.Partitioner do
+  @moduledoc """
+  Behaviour definition for partitioners, that assigns partitions for requests.
+  """
+
+  alias KafkaEx.Protocol.Produce.Request, as: ProduceRequest
+  alias KafkaEx.Protocol.Metadata.Response, as: MetadataResponse
+
+  @callback assign_partition(request :: %ProduceRequest{}, metadata :: %MetadataResponse{}) ::
+    %ProduceRequest{}
+
+  defmacro __using__(_) do
+    quote location: :keep do
+      @behaviour KafkaEx.Partitioner
+
+      @doc """
+      Returns key for given messages
+
+      Function looks for message key in messages list of {ProduceRequest}. It may return
+      either `{:ok, nil}` if no key was found, `{:ok, key}` when key was found,
+      or `{:error, message, nil | key}` when error happens while looking for the key.
+      """
+      @spec get_key(request :: %ProduceRequest{}) ::
+        {:ok, nil} | {:ok, binary} | {:error, binary, binary | nil}
+      def get_key(%ProduceRequest{messages: messages}) when is_list(messages) and length(messages) > 0 do
+        Enum.reduce(
+          messages,
+          {:ok, nil},
+          fn
+            (%{key: key}, {:ok, nil}) when is_binary(key) -> {:ok, key}
+            (%{key: key}, {:ok, key}) -> {:ok, key}
+            (%{key: other_key}, {:ok, key}) when is_binary(other_key) -> {:error, "Inconsistent keys within messages", key}
```

Comment:
If a key is set in the first message, but the second message has a key of `nil`, this 3rd clause will not catch it, and will continue as if it were set to the previously detected key.

Reply:
I think it is fine. If `key` is set only on first message (or only in 1 message), it is ok to assume it is a key for all messages.

Don't you agree?

Reply:
I agree; In that case, we probably don't need to check any beyond the first, so we can just look at the head of the list and get its key, rather than iterating over a potentially large list of messages.

Reply:
On second thought, I'm unsure. Do you know how the Java client would handle this? Does it check all messages in the set, or just the first to determine partition?

Reply:
Java client allows to send you 1 message at the time, so there is no problem with selecting the key ;).

I decided I will make `get_key` function to return key only when all messages in the set have the same key.